File: riku/core/analyzer.py
from pathlib import Path

from riku.core.git_service import GitService, LargeBlobError
from riku.core.registry import get_driver_for
from riku.core.driver import DriverDiffReport


def analyze_diff(
    repo_path: str | Path,
    commit_a: str,
    commit_b: str,
    file_path: str,
) -> DriverDiffReport:
    """
    Extrae dos revisiones de un archivo desde Git y retorna un diff semantico.

    Si el archivo no existia en commit_a o commit_b, usa b"" para ese lado
    (el driver lo interpreta como todos los elementos added/removed).
    """
    svc = GitService(repo_path)
    driver = get_driver_for(file_path)

    if driver is None:
        report = DriverDiffReport(file_type="unknown")
        report.warnings.append(f"{file_path}: no hay driver disponible para este formato.")
        return report

    content_a = _safe_get_blob(svc, commit_a, file_path)
    content_b = _safe_get_blob(svc, commit_b, file_path)

    report = driver.diff(content_a, content_b, path_hint=file_path)

    if isinstance(content_a, LargeBlobError):
        report.warnings.append(str(content_a))
    if isinstance(content_b, LargeBlobError):
        report.warnings.append(str(content_b))

    return report


def _safe_get_blob(svc: GitService, commit_ish: str, file_path: str) -> bytes:
    """
    Retorna el blob como bytes.
    - Si el archivo no existe en ese commit: retorna b"".
    - Si el blob es demasiado grande: retorna b"" y deja que el caller maneje LargeBlobError.
    """
    try:
        return svc.get_blob(commit_ish, file_path)
    except KeyError:
        return b""
    except LargeBlobError:
        raise

File: riku/src/core/analysis/analyzer.rs
use std::path::Path;

use thiserror::Error;

use crate::adapters::registry::get_driver_for;
use crate::core::analysis::blob_io;
use crate::core::domain::driver::DriverDiffReport;
use crate::core::domain::git_types::GitError;
use crate::core::domain::models::FileFormat;
use crate::core::domain::ports::GitRepository;
use crate::core::git::git_service::GitService;

#[derive(Debug, Error)]
pub enum AnalyzeError {
    #[error(transparent)]
    Git(#[from] GitError),
}

pub fn analyze_diff(
    repo_path: &Path,
    commit_a: &str,
    commit_b: &str,
    file_path: &str,
) -> Result<DriverDiffReport, AnalyzeError> {
    let svc = GitService::open(repo_path)?;
    analyze_diff_with_repo(&svc, commit_a, commit_b, file_path)
}

pub fn analyze_diff_with_repo<R: GitRepository + ?Sized>(
    repo: &R,
    commit_a: &str,
    commit_b: &str,
    file_path: &str,
) -> Result<DriverDiffReport, AnalyzeError> {
    let driver = match get_driver_for(file_path) {
        Some(driver) => driver,
        None => {
            let mut report = DriverDiffReport {
                file_type: FileFormat::Unknown,
                ..Default::default()
            };
            report.warnings.push(format!(
                "{file_path}: no hay driver disponible para este formato."
            ));
            return Ok(report);
        }
    };

    let mut warnings = Vec::new();
    let content_a = blob_io::read_blob_lenient(repo, commit_a, file_path, &mut warnings)?
        .unwrap_or_default();
    let content_b = blob_io::read_blob_lenient(repo, commit_b, file_path, &mut warnings)?
        .unwrap_or_default();

    let mut report = driver.diff(&content_a, &content_b, file_path);
    report.warnings.extend(warnings);
    Ok(report)
}
